service/controller/resource/ipam/subnet_collector.go

```diff
@@ -0,0 +1,236 @@
+}
+
+func (c *SubnetCollector) getSubnetsFromAzureConfigs(ctx context.Context) ([]net.IPNet, error) {
+	azureConfigList, err := c.g8sClient.ProviderV1alpha1().AzureConfigs(metav1.NamespaceAll).List(metav1.ListOptions{})
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	var results []net.IPNet
+	for _, ac := range azureConfigList.Items {
+		cidr := key.AzureConfigNetworkCIDR(ac)
+		if cidr == "" {
+			continue
+		}
+
+		_, n, err := net.ParseCIDR(cidr)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+
+		results = append(results, *n)
+	}
+
+	return results, nil
+}
+
+func (c *SubnetCollector) getSubnetsFromAllSubscriptions(ctx context.Context) ([]net.IPNet, error) {
+	secretsList, err := c.k8sclient.CoreV1().Secrets(metav1.NamespaceAll).List(metav1.ListOptions{
+		LabelSelector: key.OrganizationSecretsLabelSelector,
+	})
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	var doneSubscriptions []string
+	var ret []net.IPNet
+	for _, secret := range secretsList.Items {
+		clientSet, err := credential.GetAzureClientSetFromSecretName(c.k8sclient, secret.Name, secret.Namespace)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+
+		// We want to check only once per subscription.
+		if inArray(doneSubscriptions, clientSet.SubscriptionID) {
+			continue
+		}
+
+		nets, err := c.getSubnetsFromSubscription(ctx, clientSet)
+		if err != nil {
+			// We can't use this Azure credentials. Might be wrong in the Secret file.
+			// We shouldn't block the network calculation for this reason.
+			c.logger.LogCtx(ctx, "level", "warning", "message", fmt.Sprintf("Error getting used subnets for subscription %s: %s", clientSet.SubscriptionID, err))
+			continue
+		}
+
+		doneSubscriptions = append(doneSubscriptions, clientSet.SubscriptionID)
+		ret = append(ret, nets...)
+	}
+
+	return ret, nil
+}
+
+func (c *SubnetCollector) getSubnetsFromSubscription(ctx context.Context, clientSet *client.AzureClientSet) ([]net.IPNet, error) {
+	groupsClient := clientSet.GroupsClient
+	vnetClient := clientSet.VirtualNetworkClient
+
+	// Look for all resource groups that have a tag named 'GiantSwarmInstallation' with installation name as value.
+	iterator, err := groupsClient.ListComplete(ctx, fmt.Sprintf("tagName eq 'GiantSwarmInstallation' and tagValue eq '%s'", c.installationName), nil)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	var ret []net.IPNet
+
+	for iterator.NotDone() {
+		group := iterator.Value()
+
+		// Search a VNET with any of the expected names.
+		vnetCandidates := []string{
+			fmt.Sprintf("%s-VirtualNetwork", *group.Name),
+			c.installationName,
+		}
+
+		for _, vnetName := range vnetCandidates {
+			vnet, err := vnetClient.Get(ctx, *group.Name, vnetName, "")
+			if IsNotFound(err) {
+				// VNET with desired name not found, ignore this resource group.
+			} else if err != nil {
+				return nil, microerror.Mask(err)
+			} else {
+				for _, cidr := range *vnet.AddressSpace.AddressPrefixes {
```

Comment:
Nice! I think this should now also cover the node pool subnets that are added to tenant cluster vnet, right?

Reply:
yes